### PR TITLE
Enhance broadcasting and where

### PR DESCRIPTION
**Pull request by tehrengruber**
Enhances the broadcasting and where builtins to accept scalar / literal arguments.

```python
broadcast(1., (IDim,))
where(mask, a, 10)
```

### CODE REVIEW

File: src/functional/ffront/foast_passes/type_deduction.py [hunk 1]
```diff
@@ -359,22 +359,25 @@ def _visit_where(self, node: foast.Call, **kwargs) -> foast.Call:
...
             args=node.args,
             kwargs=node.kwargs,
-            type=left_type,
+            type=return_type,
             location=node.location,
         )
 
     def _visit_broadcast(self, node: foast.Call, **kwargs) -> foast.Call:
-        field_type = cast(ct.FieldType, node.args[0].type)
+        arg_type = cast(ct.FieldType | ct.ScalarType, node.args[0].type)
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/functional/ffront/foast_passes/type_deduction.py:380]: Yes, this is guaranteed by the signature declaration in `fbuiltins.py`.

File: src/functional/ffront/foast_to_itir.py [hunk 2]
```diff
@@ -75,6 +75,10 @@ def can_be_value_or_iterator(symbol_type: ct.SymbolType):
     return resulting_type_kind(symbol_type) is not TypeKind.UNKNOWN
 
 
+def is_field(expr: foast.Expr) -> bool:
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/functional/ffront/foast_to_itir.py:78]: No, as we check the type with respect to the frontend, not the type of the `expr` argument.

File: src/functional/ffront/common_types.py [hunk 3]
```diff
@@ -35,7 +35,7 @@ class SymbolType:
 class DeferredSymbolType(SymbolType):
     """Dummy used to represent a type not yet inferred."""
 
-    constraint: typing.Optional[typing.Type[SymbolType]]
+    constraint: typing.Optional[typing.Type[SymbolType] | tuple[typing.Type[SymbolType], ...]]
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/functional/ffront/common_types.py]: Cleaned up some other related things.

**Review by egparedes — APPROVED** [review]
It looks good. It can be merged as soon as CI is green.

**Pull request merged**